modules/apps/27-interchain-accounts/keeper/relay_test.go

```diff
@@ -129,105 +164,361 @@ func (suite *KeeperTestSuite) TestOnRecvPacket() {
+			func() {
+				transferPath := ibctesting.NewPath(suite.chainB, suite.chainC)
+				transferPath.EndpointA.ChannelConfig.PortID = ibctesting.TransferPort
+				transferPath.EndpointB.ChannelConfig.PortID = ibctesting.TransferPort
+
+				suite.coordinator.Setup(transferPath)
+
+				interchainAccountAddr, found := suite.chainB.GetSimApp().ICAKeeper.GetInterchainAccountAddress(suite.chainB.GetContext(), path.EndpointA.ChannelConfig.PortID)
+				suite.Require().True(found)
+
+				msg := &transfertypes.MsgTransfer{
+					SourcePort:       transferPath.EndpointA.ChannelConfig.PortID,
+					SourceChannel:    transferPath.EndpointA.ChannelID,
+					Token:            sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(100)),
+					Sender:           interchainAccountAddr,
+					Receiver:         suite.chainA.SenderAccount.GetAddress().String(),
+					TimeoutHeight:    clienttypes.NewHeight(0, 100),
+					TimeoutTimestamp: uint64(0),
+				}
+
+				data, err := types.SerializeCosmosTx(suite.chainA.GetSimApp().AppCodec(), []sdk.Msg{msg})
+				suite.Require().NoError(err)
+
+				icaPacketData := types.InterchainAccountPacketData{
+					Type: types.EXECUTE_TX,
+					Data: data,
+				}
+
+				packetData = icaPacketData.GetBytes()
+			},
+			true,
+		},
+		{
+			"cannot unmarshal interchain account packet data",
+			func() {
 				packetData = []byte{}
-			}, false,
+			},
+			false,
+		},
+		{
+			"cannot deserialize interchain account packet data messages",
+			func() {
+				data := []byte("invalid packet data")
+
+				icaPacketData := types.InterchainAccountPacketData{
+					Type: types.EXECUTE_TX,
+					Data: data,
+				}
+
+				packetData = icaPacketData.GetBytes()
+			},
+			false,
 		},
 		{
-			"Unauthorised: Interchain account not found for given source portID", func() {
-				sourcePort = "invalid-port-id"
-			}, false,
+			"invalid packet type - UNSPECIFIED",
+			func() {
+				data, err := types.SerializeCosmosTx(suite.chainA.GetSimApp().AppCodec(), []sdk.Msg{&banktypes.MsgSend{}})
+				suite.Require().NoError(err)
+
+				icaPacketData := types.InterchainAccountPacketData{
+					Type: types.UNSPECIFIED,
+					Data: data,
+				}
+
+				packetData = icaPacketData.GetBytes()
+			},
+			false,
 		},
 		{
-			"Unauthorised: Signer of message is not the interchain account associated with sourcePortID", func() {
-				// build MsgSend
-				amount, _ := sdk.ParseCoinsNormalized("100stake")
-				// Incorrect FromAddress
-				msg = &banktypes.MsgSend{FromAddress: suite.chainB.SenderAccount.GetAddress().String(), ToAddress: suite.chainB.SenderAccount.GetAddress().String(), Amount: amount}
-				// build packet data
-				data, err := types.SerializeCosmosTx(suite.chainA.Codec, []sdk.Msg{msg})
+			"unauthorised: interchain account not found for controller port ID",
+			func() {
+				path.EndpointA.ChannelConfig.PortID = "invalid-port-id"
+
+				data, err := types.SerializeCosmosTx(suite.chainA.GetSimApp().AppCodec(), []sdk.Msg{&banktypes.MsgSend{}})
```

Comment:
I think you need to populate the `&banktypes.MsgSend{}` in order to hit the intended error. I might be wrong though, but I'm guessing this fails on the message validation and not the ica auth.

Reply:
Ah sorry, I see you added a test below for that. The diff + unauthorized message threw me off.

Reply:
Yep, these two tests cover two different errors returned from `AuthenticateTx` :)